dataplane/driver.go

```diff
@@ -162,19 +162,21 @@ func StartDataplaneDriver(configParams *config.Config,
 			log.WithError(err).Warning("Unable to assign table index for wireguard")
 		}
 
-		// If wireguard is enabled, update the failsafe ports to inculde the wireguard port.
+		// If wireguard is enabled, update the failsafe ports to include the wireguard port.
 		failsafeInboundHostPorts := configParams.FailsafeInboundHostPorts
 		failsafeOutboundHostPorts := configParams.FailsafeOutboundHostPorts
 		if configParams.WireguardEnabled {
 			failsafeInboundHostPorts = make([]config.ProtoPort, len(configParams.FailsafeInboundHostPorts)+1)
 			copy(failsafeInboundHostPorts, configParams.FailsafeInboundHostPorts)
 			failsafeInboundHostPorts[len(configParams.FailsafeInboundHostPorts)] = config.ProtoPort{
+				Net: "0.0.0.0/0",
```

Comment:
This defaulting logic for the wireguard ports I think needs a rework. It looks like if we just add the ports regardless of what was set in the failsafes which worked when we were specifying just ports, but now may be too permissive. I think the logic needs a small rework to check if any of the wireguard ports have been set, and if they have not, then add a rule that allows the wireguard port over all IPs. This should allow for anyone to set their specific wireguard overrides (with the specific node IPs of our cluster) and prevent the wireguard port on all nodes to be open. WDYT?

Reply:
Yes, this makes sense!

Reply:
I've added this check in https://github.com/projectcalico/felix/pull/2646/commits/30a6abbf9dae364754729ea02fbbd204c87a3de0 but I haven't had a chance to test this change. I'm also struggling to find a place where we can test this easily, as it stands that dataplane driver doesn't have a test suite that one can easily utilise.

Reply:
Ah yeah, this one is a little difficult given how we haven't really created a set of dataplane tests for this. Probably the easiest way to quickly verify is to add something like a `Debug` statement or something where the failsafes get applied and then build a new node image with these felix changes. Then you can run the changed node in a cluster quickly, while passing the appropriate configuration, and verify if your debug statement shows the appropriate failsafes when you check the pod logs.

The proper solution is honestly a little more work than I would want for a change as small as this, which would be to create the suite of dataplane tests but even that wouldn't be very easy for this test since there's no easy way to inspect the config that the dataplane is created with without writing a bunch of new methods just to get that information.

Reply:
Hi @mgleung I've tested this within our cluster by using debug logs as you suggested.

Default wireguard config, if a more specific entry for wireguard wasn't found in the failsafe config
```
2021-03-04 14:34:36.501 [DEBUG][52] felix/driver.go 184: Adding permissive FailsafeInboundHostPorts for wireguard
2021-03-04 14:34:36.501 [DEBUG][52] felix/driver.go 207: Adding permissive FailsafeOutboundHostPorts for wireguard

-A cali-failsafe-in -p udp -m comment --comment "cali:sePiEYnWQoE34rxm" -m multiport --dports 51820 -j ACCEPT
-A cali-failsafe-out -p udp -m comment --comment "cali:cKMbiSScdOAMPjj7" -m multiport --dports 51820 -j ACCEPT
```

When a specific entry for wireguard is found in the failsafe config
```
2021-03-04 14:41:30.111 [DEBUG][52] felix/driver.go 176: FailsafeInboundHostPorts is already configured for wireguard net="172.0.0.0/8" port=0xca6c protocol="udp"
2021-03-04 14:41:30.111 [DEBUG][52] felix/driver.go 199: FailsafeOutboundHostPorts is already configured for wireguard net="172.0.0.0/8" port=0xca6c protocol="udp"

-A cali-failsafe-in -s 172.0.0.0/8 -p udp -m comment --comment "cali:KoXgB1rx1y51vKyk" -m multiport --dports 51820 -j ACCEPT
-A cali-failsafe-out -d 172.0.0.0/8 -p udp -m comment --comment "cali:2KMZG0hyUESpYCVf" -m multiport --dports 51820 -j ACCEPT